### PR TITLE
Add since information to properties

### LINKED ISSUE #1117: Add since information to properties
https://issues.apache.org/jira/browse/ACCUMULO-4592

### PR DESCRIPTION
**Pull request by foster33**
Fixes #1117 

In this PR the following has been added:

- A new enum constructor parameter in `Property.java` that provides `since` information about when the property was first introduced
- Updated the information provided in `server-properties.md`
  - Added `deprecated since` information where applicable
  - Added `replaced by` information where applicable
  - Added `available since` information where applicable
- Provided all existing properties with their `availableSince` information
- Removed redundant information within existing property descriptions that are now found elsewhere

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (Manno15) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java:83]: This comment is probably unnecessary now with the since info below.

Comment (foster33) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:83]: I did notice that in a couple places. I left it alone because I did not know if it was important / useful to keep it in the javadoc. I will remove this if it is not needed anymore.

**Review by Manno15 — APPROVED** [review]
Code changes look good to me, though I did not check if the since dates were accurate or not.

**Comment by foster33** [comment]
> Code changes look good to me, though I did not check if the since dates were accurate or not.

I did make some decisions on some versions that the properties have. For example, for the `manager` properties, I decided to use the version number of the property before the name change. I thought this would be more useful than simply putting 2.1.0 for all of them. If this was wrong to think let me know and I can go back through them and change the version number.

**Comment by foster33** [comment]
I had also thought of another addition that may be useful. When it mentions that a property is "replaced by X" it may be useful to link to the property that is replacing the old one. I went ahead and made the changes for this and will wait for feedback as to if this would be useful / if I made the changes correctly.

**Comment by foster33** [comment]
Selection_011

Here is an example of what some of the changes may look like.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (DomGarguilo) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java:1336]: I think this is causing the test failure
```suggestion
  private String deprecatedSince;
```

File: core/src/main/java/org/apache/accumulo/core/conf/ConfigurationDocGen.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/conf/ConfigurationDocGen.java:89]: This code looks good, but I think we want to show "Available since" regardless of if it's deprecated. The above logic will only show it if it is not deprecated.

Also, can you work in line breaks (`<br>`), so that "Available since", "Deprecated since", and "Replaced by" show up on separate lines, above the description?

File: core/src/main/java/org/apache/accumulo/core/conf/ConfigurationDocGen.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/conf/ConfigurationDocGen.java:104]: Same comment here as above for the prefix types.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:83]: Feel free to get rid of them.

**Review by ctubbsii — COMMENTED** [review]
Implementation looks good overall. My main comments are regarding style (newlines) in the output. Great job.

I haven't checked all the values for the "availableSince", and I'm curious how you populated those, but trust that they're correct. Perhaps somebody will be able to verify before merging.

As for the 2.1 manager properties, I think it was a good idea to track the since values from before they were automatically renamed. However, I'm not sure which is more useful to display. Since the availableSince is an arbitrary String, you could have something like "2.1.0; since 1.3.6 as 'oldname'" or similar. That might get a bit redundant in the generated view, though, so you may wish to just put some basic explanation in the description of the manager prefix property.

**Comment by foster33** [comment]
> I haven't checked all the values for the "availableSince", and I'm curious how you populated those, but trust that they're correct. Perhaps somebody will be able to verify before merging.

I had gone through each of the tags available on github starting at 1.3.5 and continued until the current state. I looked at `Property.java` each time and compared the file to the previous version tag and would mark down the version number of any new properties.

**Comment by foster33** [comment]
Selection_012

Here is an updated example of the output after the recent commit.

**Comment by ctubbsii** [comment]
> I had gone through each of the tags available on github starting at 1.3.5

Wow, thanks for being so thorough! The output looks nice, too. I'm going to take a look at the overall result in the generated output to see if I have any additional feedback, but as far as I can tell, this looks great! I think these will be very useful to users.

**Review by ctubbsii — APPROVED** [review]
@foster33 I added two commits to your PR. The first slightly improves some wording and formatting. The second converts the ternary statements to if-else statements, which should make the whole thing easier to maintain (if we had done that before, your changes would have probably been easier to implement).

With these additional changes, I'm ready to merge this in, but since I added two new commits, I wanted to double check back with you, to see if you're okay with these changes, or had any other proposed changes, before merging.

**Comment by DomGarguilo** [comment]
I'm almost done double checking the since tags. Should be finished early tomorrow.

**Comment by foster33** [comment]
> With these additional changes, I'm ready to merge this in, but since I added two new commits, I wanted to double check back with you, to see if you're okay with these changes, or had any other proposed changes, before merging.

These changes look good to me!

**Comment by DomGarguilo** [comment]
I went through and double checked that the correct since tags were present. Below are my findings. Please double check what I found before making changes.

MASTER_RECOVERY_DELAY seen at 1.5.0 2db5ce6186e32c451328154b024951cc5090505f
GENERAL_KERBEROS_RENEWAL_PERIOD marked 1.6.5 seen at 1.7.1 3aa9d307d7405d8d98c02dcc11c94cfe07ea2a14
TSERV_UPDATE_SESSION_MAXIDLE marked 1.6.6 seen at 1.6.5 46ad8368e160c56c03571b467f8ae603c50992f6
TABLE_FILE_SUMMARY_MAX_SIZE marked at 1.4.0 seen at 2.0.0 94cdcc4d3f0a8ccf95894f206cb71e6117f4e51d
MANAGER_RENAME_THREADS marked as 1.8.0 seen at 1.7.0 5cbcffbb4ce08325af6451eb2941ece6c3886598 
MANAGER_BULK_THREADPOOL_TIMEOUT marked 1.4.3 seen at (2.1.0?) 352d2cd050411233148fd476531d93297b98d308 originally 'MASTER_BULK_THREADPOOL_ALLOW_TIMEOUT'
TSERV_UPDATE_SESSION_MAXIDLE marked 1.6.6 seen at 1.6.5 46ad8368e160c56c03571b467f8ae603c50992f6
TSERV_MINTHREADS_TIMEOUT marked 1.4.3 seen at 2.1.0? 352d2cd050411233148fd476531d93297b98d308 originally 'MASTER_MINTHREADS_ALLOW_TIMEOUT'
TABLE_FILE_SUMMARY_MAX_SIZE marked 1.4.0 seen at 2.0.0 94cdcc4d3f0a8ccf95894f206cb71e6117f4e51d

**Comment by foster33** [comment]
> Please double check what I found before making changes.

I went through and checked your proposed corrections and everything looked good besides:

GENERAL_KERBEROS_RENEWAL_PERIOD -- found in rel/1.6.5 not 1.7.1
MANAGER_RENAME_THREADS -- we are both incorrect. After looking into it further, it was first seen in 2.1.0

I will go ahead and make these corrections in my next commit. Thank you for double checking!

**Comment by DomGarguilo** [comment]
> GENERAL_KERBEROS_RENEWAL_PERIOD -- found in rel/1.6.5 not 1.7.1
> MANAGER_RENAME_THREADS -- we are both incorrect. After looking into it further, it was first seen in 2.1.0

Looks like you are correct here.

**Review by DomGarguilo — APPROVED** [review]
Looks good to me once the tests pass.

**Review by Manno15 — APPROVED** [review]

**Pull request merged**
